spec/mailers/bulk_upload_mailer_spec.rb

```diff
@@ -109,13 +109,8 @@
   end
 
   describe "#send_correct_and_upload_again_mail" do
-    context "when 2 columns with errors" do
-      before do
-        create(:bulk_upload_error, bulk_upload:, col: "A")
-        create(:bulk_upload_error, bulk_upload:, col: "B")
-      end
-
-      it "sends correctly formed email with A, B" do
+    context "when there are no errors" do
+      it "sends correctly formed email" do
         expect(notify_client).to receive(:send_email).with(
```

Comment:
Does this expect line get called after the mail has been sent (line 127)? If so, why is it before it in the file?

Reply:
- if you want it read so its after you have to use `have_received` see https://rubydoc.info/github/rspec/rspec-mocks/RSpec%2FMocks%2FExampleMethods:have_received
- this is just how the `receive` part of the DSL works. you set up the expectation before, call the thing. this way it knows if the expectation happened or not.
- `have_received` works the same way except you write it afterwards
- i don't actually know how it works under the hood but i would guess it setups all the expectations up first then calls the code so it knows if mocks received messages or not